### PR TITLE
[flang][OpenMP] Lower `target .. private(..)` to `omp.private` ops

### PR DESCRIPTION
**Pull request by ergawy**
Extends delayed privatization support to `taraget .. private(..)`. With this PR, `private` is support for `target` **only** is delayed privatization mode.

### CODE REVIEW

**Comment by ergawy** [comment]
Ping 🔔. Please try to take a look on this PR!

File: flang/lib/Lower/OpenMP/OpenMP.cpp [hunk 1]
```diff
@@ -758,15 +758,33 @@ genBodyOfTargetOp(lower::AbstractConverter &converter, lower::SymMap &symTable,
...
+  llvm::SmallVector<mlir::Location> allRegionArgLocs;
+  allRegionArgTypes.reserve(mapSymTypes.size() +
+                            targetOp.getPrivateVars().size());
+  llvm::transform(mapSymLocs, std::back_inserter(allRegionArgLocs),
+                  [](mlir::Location l) { return l; });
+  llvm::transform(privateVars, std::back_inserter(allRegionArgLocs),
+                  [](mlir::Value v) { return v.getLoc(); });
+
+  auto *regionBlock = firOpBuilder.createBlock(&region, {}, allRegionArgTypes,
+                                               allRegionArgLocs);
```
Comment (kiranchandramohan) [inline flang/lib/Lower/OpenMP/OpenMP.cpp:787]: Add a comment for this block of code. Is it possible to share any of it with the parallel operation? Or is it not possible due to the isolated from above nature of target?

**Review by kiranchandramohan — COMMENTED** [review]
Could you add a few more tests?
-> With a default private.
-> With three different types of privates.

Did you get a chance to discuss with @mjklemm about how privatisation works for tasks and since target has some task like properties whether we need something more than inlining?

**Comment by ergawy** [comment]
Thanks for taking a look @kiranchandramohan.

> Could you add a few more tests?
> -> With a default private.

I might have missed something, but `target` does not support the `default` clause, right?

> -> With three different types of privates.

You mean different types of variables; e.g. primitive, allocatable, pointer, array, etc? Or something else?

> Did you get a chance to discuss with @mjklemm about how privatisation works for tasks and since target has some task like properties whether we need something more than inlining?

Not yet. Will do that. (cc @mjklemm & @pranavb-ca since we discussed `target` support in MLIR -> LLVM translation layer)

**Comment by bhandarkar-pranav** [comment]
> Could you add a few more tests? -> With a default private. -> With three different types of privates.
> 
> Did you get a chance to discuss with @mjklemm about how privatisation works for tasks and since target has some task like properties whether we need something more than inlining?

Yes, we'll need a way to send the privates via the target task to the target kernel so that `firstprivate` variables can be initialized. The way clang does this is by tacking on the private variables at the end of the `task_t` structure in the call to `kmpc_omp_task_alloc`. For now, this is a TODO item for both, [Tasks](https://github.com/llvm/llvm-project/blob/main/llvm/lib/Frontend/OpenMP/OMPIRBuilder.cpp#L1790) and [Target](https://github.com/bhandarkar-pranav/llvm-project/blob/ompirbuilder/target_task_codegen/llvm/lib/Frontend/OpenMP/OMPIRBuilder.cpp#L5529) in the `OpenMPIRBuilder`. These values can then be picked up by the inlined code (generated out of the privatizers by way of a `PrivCB` callback) in the target kernel.

File: flang/lib/Lower/OpenMP/OpenMP.cpp (hunk 1 above)
Comment (ergawy) [reply flang/lib/Lower/OpenMP/OpenMP.cpp]: Abstracted the common pieces into one template and commented it. Let me know if you don't like how I handled this or if it is too abstract and/or noisy.

**Comment by kiranchandramohan** [comment]
> I might have missed something, but target does not support the default clause, right?

Sorry, i completely missed that.

> You mean different types of variables; e.g. primitive, allocatable, pointer, array, etc? Or something else?

I specifically meant types here (integer, real, complex, character). But you can include arrays, allocatable etc. Just to ensure we are correctly handling the case when more than one type is present. I suggest also trying out multiple private clauses.

> > Could you add a few more tests? -> With a default private. -> With three different types of privates.
> > Did you get a chance to discuss with @mjklemm about how privatisation works for tasks and since target has some task like properties whether we need something more than inlining?
> 
> Yes, we'll need a way to send the privates via the target task to the target kernel so that `firstprivate` variables can be initialized. The way clang does this is by tacking on the private variables at the end of the `task_t` structure in the call to `kmpc_omp_task_alloc`. For now, this is a TODO item for both, [Tasks](https://github.com/llvm/llvm-project/blob/main/llvm/lib/Frontend/OpenMP/OMPIRBuilder.cpp#L1790) and [Target](https://github.com/bhandarkar-pranav/llvm-project/blob/ompirbuilder/target_task_codegen/llvm/lib/Frontend/OpenMP/OMPIRBuilder.cpp#L5529) in the `OpenMPIRBuilder`. These values can then be picked up by the inlined code (generated out of the privatizers by way of a `PrivCB` callback) in the target kernel.

What I am worried about is whether we want a different style entry in the PrivateClause Operation so that the firstprivate variable can be laid out properly in the privatisation part of the `task_t` structure.
Would it make sense to add a TODO/`Not Yet Implemented Error` for all target directives for firstprivate clauses that can be deferred?

Separately, you might need something like @luporl's patch for default firstprivate behaviour. https://github.com/llvm/llvm-project/pull/85989

**Comment by bhandarkar-pranav** [comment]
> What I am worried about is whether we want a different style entry in the PrivateClause Operation so that the firstprivate variable can be laid out properly in the privatisation part of the `task_t` structure.

I am afraid I dont really understands what you mean here, specifically when you say "different style entry in PrivateClause Operation". Could you please elaborate?

> Would it make sense to add a TODO/`Not Yet Implemented Error` for all target directives for firstprivate clauses that can be deferred?
I think there's no harm in separating the translation of `private` from `firstprivate` on target constructs by deferring the `firstprivate` work. (Implementing a TODO)
> Separately, you might need something like @luporl's patch for default firstprivate behaviour. #85989
Thank you for bringing to my notice.

**Comment by kiranchandramohan** [comment]
> > What I am worried about is whether we want a different style entry in the PrivateClause Operation so that the firstprivate variable can be laid out properly in the privatisation part of the `task_t` structure.
> 
> I am afraid I dont really understands what you mean here, specifically when you say "different style entry in PrivateClause Operation". Could you please elaborate?

To layout a copy of the original variable in the task datastructure, I am assuming you will need some kind of size, offset and alignment information. With a PrivateClause operation we have a sequence of MLIR operations. It is not clear whether we can deduce the necessary information for laying out a firstprivate variable in the task datastructure.

**Comment by ergawy** [comment]
> Would it make sense to add a TODO/`Not Yet Implemented Error` for all target directives for firstprivate clauses that can be deferred?

`firstprivate` is currrently still a TODO for `target`:
```c++
  cp.processTODO<clause::Allocate, clause::Defaultmap, clause::Firstprivate,
                 clause::InReduction, clause::Reduction,
                 clause::UsesAllocators>(loc,
                                         llvm::omp::Directive::OMPD_target);
```

This PR is only for `private`. Let me know if I misunderstood what you mean.

I also added a more extensive test.

File: flang/lib/Lower/OpenMP/OpenMP.cpp [hunk 2]
```diff
@@ -456,6 +456,33 @@ markDeclareTarget(mlir::Operation *op, lower::AbstractConverter &converter,
...
 }
 
+/// For an operation that takes `omp.private` values as region args, this util
+/// merges the private vars info into the region arguments list.
+///
+/// \tparam OMPOP  - the OpenMP op that takes `omp.private` inputs.
+/// \tparam InfoTy - the type of private info we want to merge; e.g. mlir::Type
+/// or mlir::Location fields of the private var list.
+///
+/// \param [in] op                 - the op accpeting `omp.private` inputs.
```
Comment (kiranchandramohan) [inline flang/lib/Lower/OpenMP/OpenMP.cpp:466]:
```suggestion
/// \param [in] op                 - the op accepting `omp.private` inputs.
```

**Review by kiranchandramohan — APPROVED** [review]
LG.

**Pull request merged**